Re-add missing public APIs in NpgsqlCommand and NpgsqlTransaction

baronfel opened the pull request:
Fixes #2607 by adding the two APIs that were altered there. Prospectively opening this MR to get checks started.

baronfel commented on src/Npgsql/NpgsqlCommand.cs:
```diff
@@ -1052,7 +1060,7 @@ public new NpgsqlDataReader ExecuteReader(CommandBehavior behavior = CommandBeha
 
         /// <summary>
-        /// An asynchronous version of <see cref="ExecuteReader"/>, which executes
+        /// An asynchronous version of <see cref="ExecuteReader()"/>, which executes
```
This change was because the overload was ambiguous between the no-parameter and default-parameter members of the same name.

YohDeadfall commented:
The `dev` branch is for version 5, so you need to retarget the PR to `release/4.1`.

baronfel commented:
Done, thanks for the reminder.

baronfel commented:
I used [SynVer.Lib](https://github.com/fsprojects/SyntacticVersioning/tree/master/src/SynVer.Lib) and the following F# script to diff the surface area of the two dlls:

```fsharp
#I "/usr/local/share/dotnet/shared/Microsoft.NETCore.App/3.0.0-preview9-19423-09"
#r "netstandard"
#r "/Users/chethusk/.nuget/packages/synver.lib/0.0.8/lib/netstandard2.0/SynVer.Lib.dll"

open System.Reflection
open SynVer

let previewPath = "/Users/chethusk/.nuget/packages/npgsql/4.1.0-preview2/lib/netcoreapp3.0/Npgsql.dll"
let oldPath = "/Users/chethusk/.nuget/packages/npgsql/4.0.9/lib/netstandard2.0/Npgsql.dll"

let preview = Assembly.LoadFile previewPath |> SurfaceArea.ofAssembly
let old = Assembly.LoadFile oldPath |> SurfaceArea.ofAssembly

let diffs = SurfaceArea.diff old preview

System.IO.File.WriteAllLines("/Users/chethusk/oss/foobar/diff.txt", diffs)
```

The full report is attached, I haven't gone over it 100% yet.

[diff.txt](https://github.com/npgsql/npgsql/files/3587092/diff.txt)

baronfel commented:
From that report, the list of missing or changed members:

* Npgsql.BackendMessages.FieldDescription: default constructor gone
* Npgsql.INpgsqlDatabaseInfoFactory.Load: timeout parameter changed namespace
* Npgsql.NpgsqlBinaryImporter.Complete: `void` return changed to `uint64`
* Npgsql.NpgsqlCommand.ExecuteReader: already noticed/reverted
* Npgsql.NpgsqlConnection.WaitAsync: `unit -> Task` overload removed
* Npgsql.NpgsqlDataReader: changed to concrete class
* Npgsql.NpgsqlDataReader.GetStreamAsync: no-cancellation-token overload removed
* Npgsql.NpgsqlDataReader.GetTextReaderAsync: no-cancellation-token overload removed
* Npgsql.NpgsqlLargeObjectStream.GetLengthAsync: no-cancellation-token overload removed
* Npgsql.NpgsqlStatement.Rows: uint32 => uint64
* Npgsql.NpgsqlTimeout: type removed
* Npgsql.NpgsqlTransaction.IsCompleted: removed (already addressed)
* Npgsql.PostgresException: default constructor gone
* Npgsql.PostgresNotice: default constructor gone
* Npgsql.TypeHandlers.ArrayHandler.CreateArrayHandler: parameter type change from PostgresType to PostgresArrayType
* Npgsql.TypeHandlers.ArrayHandler.CreateRangeHandler: parameter type change from PostgresType to PostgresArrayType
* Npgsql.TypeHandlers.ArrayHandler<'TELEMENT>: new parameter added to constructor
* Npgsql.TypeHandlers.ByteaHandler: default constructor gone
* Npgsql.TypeHandlers.IMappedCompositeTypeHandlerFactory: type removed
* Npgsql.TypeHandlers.JsonbHandler: type removed
* Npgsql.TypeHandlers.RangeHandler<'TELEMENT>: new parameter added for range postgres type
* Npgsql.TypeHandlers.RangeHandler<'TELEMENT>.CreateArrayHandler: parameter changed from PostgresType to PostgresArrayType
* Npgsql.TypeHandling.NpgsqlSafeReadException: constructor parameter name changed
* Npgsql.TypeHandling.NpgsqlSimpleTypeHandlerWithPsv<'TDEFAULT,'TPSV>: constructor parameter type changed from PostgresType to PostgresArrayType
* Npgsql.TypeHandling.NpgsqlTypeHandler.CreateArrayHandler: parameter type changed from PostgresType to PostgresArrayType
* Npgsql.TypeHandling.NpgsqlTypeHandler.CreateRangeHandler: parameter type changed from PostgresType to PostgresArrayType
* Npgsql.TypeHandling.NpgsqlTypeHandler<'TDEFAULT>.CreateArrayHandler: parameter type changed from PostgresType to PostgresArrayType
* Npgsql.TypeHandling.NpgsqlTypeHandler<'TDEFAULT>.CreateRangeHandler: parameter type changed from PostgresType to PostgresArrayType
* Npgsql.TypeHandling.NpgsqlTypeHandlerFactory: type removed
* Npgsql.TypeMapping.NpgsqlTypeMapping.TypeHandlerFactory: property type changed from NpgsqlTypeHandlerFactory to INpgsqlTypeHandlerFactory
* Npgsql.TypeHandling.NpgsqlTypeHandlerFactory<'TDEFAULT>: changed base class to object
* Npgsql.TypeMapping.NpgsqlTypeMappingBuilder.TypeHandlerFactory: property type changed from NpgsqlTypeHandlerFactory to INpgsqlTypeHandlerFactory
* NpgsqlTypes.NpgsqlTsVector: list-of-lexemes constructor removed

roji commented:
No problem for bringing back parameter-less ExecuteReader, but NpgsqlTransaction.IsCompleted is a bit complicated... this was done as part of #2416, where we started recycling the NpgsqlTransaction instance. The logic of removing/hiding IsCompleted was that since the transaction instance will be revived, IsCompleted will switch from true to false, leading to potential confusion/dangerous behavior. Therefore, it is up to the user to track whether a transaction is in progress or not (which is the way ADO.NET works - IsCompleted has always been an Npgsql extension).

I'll have to investigate this a bit. Currently it is not required to actually dispose an NpgsqlTransaction before starting a new transaction - the only thing that matters is whether the transaction has completed on the PostgreSQL side. I'm not sure that's the desirable behavior. Depending on that, it may still make sense to have a public IsCompleted, with a contract that once the instance has been disposed, all bets are off. Give me a couple days to look at this.

roji commented:
@baronfel I've reverted the transaction recycling change in #2617, which also brings back the missing IsCompleted. Can you please remove that part from this PR?

roji commented:
Thanks @baronfel!

baronfel commented:
Should I log an issue with the remaining changed members to be reverted in the release/4.1 branch? I know that the TypeHandling changes have been somewhat changed, I'm speaking specifically of

```
Npgsql.BackendMessages.FieldDescription: default constructor gone
Npgsql.INpgsqlDatabaseInfoFactory.Load: timeout parameter changed namespace
Npgsql.NpgsqlBinaryImporter.Complete: void return changed to uint64
Npgsql.NpgsqlCommand.ExecuteReader: already noticed/reverted
Npgsql.NpgsqlConnection.WaitAsync: unit -> Task overload removed
Npgsql.NpgsqlDataReader: changed to concrete class
Npgsql.NpgsqlDataReader.GetStreamAsync: no-cancellation-token overload removed
Npgsql.NpgsqlDataReader.GetTextReaderAsync: no-cancellation-token overload removed
Npgsql.NpgsqlLargeObjectStream.GetLengthAsync: no-cancellation-token overload removed
Npgsql.NpgsqlStatement.Rows: uint32 => uint64
Npgsql.NpgsqlTimeout: type removed
Npgsql.NpgsqlTransaction.IsCompleted: removed (already addressed)
Npgsql.PostgresException: default constructor gone
Npgsql.PostgresNotice: default constructor gone
```

roji commented:
@baronfel yes please! It makes sense to add back the no-cancellation-token overloads, but it looks like we may keep most of the other changes listed.

BTW which tool are you using to detect these changes?

baronfel commented:
I'm using SynVer (per https://github.com/npgsql/npgsql/pull/2608#issuecomment-529151717) and a short script. There's also a standalone windows executable for it, and I'm working on an PR for the tool to allow for usage as a dotnet SDK global tool, so you could do `synver <source dll> <target dll>` to get the same output. Along with potential enhancements like a structured json output, etc.
